src/plugins/reference/README.md

```diff
@@ -0,0 +1,128 @@
+
+If a key has the metakey `check/reference` with value `single`, the plugin reads the value
+of the key and produces an error, if the value is not a valid reference.
+
+If a key has metakey `check/reference` with value `recursive`, the plugin constructs a 
+reference graph starting with the marked key. (for the exact construction see below) 
+In this reference graph each node corresponds to a key in the KDB, while each edge 
+represents a reference between to keys. The plugin will produce an error, if this graph
+is not a directed acyclic graph or contains any invalid references.
+
+Additionally to this basic functionality the plugin allows one to restrict the set of valid
+references using the `check/reference/restrict` metakey. The exact mechanism for this is 
+described below.
+
+### Resolution of references
+If the key `A` is interpreted as a reference by the plugin, it has to be an array. Each of 
+the array elements must contain a so called reference value. This reference value will be 
+resolved into a keyname, which will then be validated. The resolution goes as follows:
+
+* If the reference value starts with `/`, it is treated as absolute and interpreted as a 
+  keyname directly.
+* Otherwise the reference value will be interpreted as relative to the **parent** of `A`.
+  That means a reference which does not contain a `/` will also was refer to a sibling of `A`,
+  the reference array key.
+* `.` and `..` will be treated the same way as in UNIX paths. If these are used in an absolute
+  reference, the plugin will emit a warning, because their use in this case is unnecessary and 
+  might be a mistake. The same is true, if `.` is used anywhere, but at the start of the 
+  reference.
+
+### Construction of the reference graph
+
+The process starts with a key `X`, which has the metakey `check/reference` set to the value
+`recursive`. This key `X` must be an array key. The basename of `X` will be called the `refname`.
+For each element in the array run the following process:
+
+1. Interpret the current key `K`s value as a referemce and throw an error, if the reference
+   is not valid.
+2. If there is no node for `K` in the reference graph create one. If there is no node for 
+   `R` in the reference graph create one. Insert an edge from `K` to `R` into the graph.
+3. Find a key `K1` directly below `R` that has the `refname` as its basename. Interpret this 
+   key as an array, if found. For each array element, start the process recursively from 1 
+   and then continue with 4.
+4. Find any key `K2` directly below `R` that has the metakey `check/reference` set to 
+   `alternative`. Interpret any such key as an array. For each array element, start the 
+   process recursively from 1, but this time using the basename of `K2` (the of the array)
+   as the `refname`.
+
+Once the whole process is finished, we will have a graph of the whole reference structure 
+stemming from `X`. This graph can then be check for acyclicity.
```

Comment:
the same is needed for the fallback/override links. Maybe do this in the same plugin? Or in a separate plugin reusing the graph code?

Reply:
My plan was to represent the graph as a separate KeySet. So I think it would be easiest to just copy the code to wherever fallback/override links are currently processed.

Reply:
Maybe we can also add it to libmeta, it seems to fit to the topological sort. Or we move elektraSortTopology together with your function to libgraph?